Disallow fallback kernel in FakeTensor by default

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97715

tugsbayasgalan committed:
Disable allow_fallback_kernel for custom ops

[ghstack-poisoned]

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1233,14 +1233,20 @@ def dispatch(self, func, types, args=(), kwargs=None):
+            if not self.allow_fallback_kernels:
+                return False
+            # should't fallback on custom ops
+            return func.namespace in ["prim", "aten"]
```
you mean prims, not prim. Also, we shouldn't allow fallback on prims anyway, they should ALWAYS have meta by construction

ezyang reviewed (approved):
looks fine but please fix the glaring bug

eellison reviewed (commented):
What is the advantage of doing this instead of changing the default of `allow_fallback_kernels` to False ? Now there is no way for a user to override this behavior.

ezyang commented:
> What is the advantage of doing this instead of changing the default of allow_fallback_kernels to False ? Now there is no way for a user to override this behavior.

It's a question of defaults. By doing it this way, we guarantee that we don't IMA on indexy custom ops.

eellison commented:
Yea, understood - but why don't we just switch the default to False then ? Instead of this proposed change

ezyang commented:
I think I proposed this to you and you didn't like it because it would have reduced fake tensor coverage. For the record, I ALSO support changing this default.

eellison commented:
Ugh I think I prefer that change to this one. but i'll let you make the final call.

zou3519 reviewed (approved):
LGTM modulo Ed's comment

tugsbayasgalan committed:
Update on "Disable allow_fallback_kernel for custom ops"


For custom ops, it is better to be explicitly ask user to provide Meta implementation instead of implicitly using CPU. 


[ghstack-poisoned]

tugsbayasgalan commented:
Going to set allow_fallback_kernel to False by default.

tugsbayasgalan committed:
Update on "Disallow fallback kernel in FakeTensor by default"




[ghstack-poisoned]

ezyang commented:
But I guess we may have to add all of the missing metas now up front :(

tugsbayasgalan committed:
Update on "Disallow fallback kernel in FakeTensor by default"




[ghstack-poisoned]
